### PR TITLE
Google cloud functions

### PR DESCRIPTION
**Pull request by abaumgarten**
#Status
- [x] Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/29106

## Description
Project_id is still an optional parameter. If empty, get it from credentials_json.
When region is empty (optional param), the default region is  All regions as should be

## Minimum version of Demisto
- [x] 5.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 1]
```diff
@@ -169,15 +171,24 @@ def format_parameters(parameters: str) -> str:
...
+        # from Google API : If you want to list functions in all locations, use "-" in place of a location
+        return "-"
+    return region
+
+
 def main():
     credentials_json = json.loads(demisto.params().get('credentials_json', {}))
     project = demisto.params().get('project_id', '')
     region = demisto.params().get('region')
+    region = set_default_region(region)
```
Comment (teizenman) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:185]: You can use the following for default values:
```suggestion
    region = demisto.params().get('region', '-')
```
Add the comment above it and youre good

**Review by teizenman — CHANGES_REQUESTED** [review]
I guess that you have checked that `project_id` is mandatory right?

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py (hunk 1 above)
Comment (abaumgarten) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:185]: I did so but when I ran on Demisto region left None.. So I added this function to be on the safe side

**Review by glicht — CHANGES_REQUESTED** [review]
Let's make project_id optional and get the value from the key file. See my comment at: https://github.com/demisto/etc/issues/29106#issuecomment-707766790

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 2]
```diff
@@ -169,15 +171,37 @@ def format_parameters(parameters: str) -> str:
...
+    if project is None:
+        project = credentials_json["project_id"]
+    return project
+
+
 def main():
     credentials_json = json.loads(demisto.params().get('credentials_json', {}))
-    project = demisto.params().get('project_id', '')
+
+    # when project_id is empty, get it from credentials_json
```
Comment (teizenman) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:190]: Place this comment under the `set_default_project_id` function

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 3]
```diff
@@ -169,15 +171,37 @@ def format_parameters(parameters: str) -> str:
...
+    project = demisto.params().get('project_id')
+    project = set_default_project_id(project, credentials_json)
+
+    # when region is empty, set it to '-' meaning all regions
```
Comment (teizenman) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:194]: Place this comment under `set_default_region` function

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 4]
```diff
@@ -169,15 +171,37 @@ def format_parameters(parameters: str) -> str:
...
 
 
+def set_default_region(region):
+    if region is None:
+        # from Google API : If you want to list functions in all locations, use "-" in place of a location
+        return "-"
+    return region
+
+
+def set_default_project_id(project, credentials_json):
```
Comment (teizenman) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:181]: Please add UT for this function

**Review by teizenman — CHANGES_REQUESTED** [review]

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py (hunk 2 above)
Comment (abaumgarten) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py]: Done

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py (hunk 3 above)
Comment (abaumgarten) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py]: Done

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py (hunk 4 above)
Comment (abaumgarten) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:181]: Done

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions_test.py [hunk 5]
```diff
@@ -1,4 +1,16 @@
...
+from GoogleCloudFunctions import set_default_project_id
+
+
+@pytest.mark.parametrize('project, credentials_json, expected_output', [
+    ("some-project-id", {"credentials_json": {"type": "service_account", "project_id": "some-project-id"}},
+     "some-project-id"),
+    (None, {"credentials_json": {"type": "service_account", "project_id": "some-project-id"}}, "some-project-id"),
+    ("some-project-id", {"credentials_json": {"type": "service_account"}}, "some-project-id"),
+])
+def test_set_default_project_id(project, credentials_json, expected_output):
```
Comment (teizenman) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions_test.py:11]: Please add the edge case where you dont have both, catch the exception

**Review by teizenman — CHANGES_REQUESTED** [review]

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions_test.py (hunk 5 above)
Comment (abaumgarten) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions_test.py]: Done

**Review by teizenman — APPROVED** [review]

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 6]
```diff
@@ -169,15 +171,38 @@ def format_parameters(parameters: str) -> str:
     return json.dumps(dict(pairs))
 
 
+def set_default_region(region):
+    # when region is empty, set it to '-' meaning all regions
+    # note : demisto.params().get('region','-') did not worked on Demisto
+    if region is None:
```
Comment (glicht) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:177]: Handle also the case of empty string:
```suggestion
    if not region:
```

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 7]
```diff
@@ -169,15 +171,38 @@ def format_parameters(parameters: str) -> str:
...
+        # from Google API : If you want to list functions in all locations, use "-" in place of a location
+        return "-"
+    return region
+
+
+def set_default_project_id(project, credentials_json):
+    if project is None:
```
Comment (glicht) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:184]: Handle also the case of empty string
```suggestion
    if not project:
```

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 8]
```diff
@@ -169,15 +171,38 @@ def format_parameters(parameters: str) -> str:
     return json.dumps(dict(pairs))
 
 
+def set_default_region(region):
```
Comment (glicht) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:174]: I would rename this function to `get_...` or `resolve_...`. It is not setting anything.

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 9]
```diff
@@ -169,15 +171,38 @@ def format_parameters(parameters: str) -> str:
...
+    # when region is empty, set it to '-' meaning all regions
+    # note : demisto.params().get('region','-') did not worked on Demisto
+    if region is None:
+        # from Google API : If you want to list functions in all locations, use "-" in place of a location
+        return "-"
+    return region
+
+
+def set_default_project_id(project, credentials_json):
```
Comment (glicht) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:183]: I would rename this function to `get_...` or `resolve_...`. It is not setting anything.

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py (hunk 8 above)
Comment (abaumgarten) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py]: Done

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py (hunk 6 above)
Comment (abaumgarten) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py]: Done

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py (hunk 9 above)
Comment (abaumgarten) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py]: Done

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py (hunk 7 above)
Comment (abaumgarten) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py]: Done

**Review by glicht — APPROVED** [review]

**Pull request merged**